include/kernels/madwf_transfer.cuh

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <color_spinor_field.h>
+#include <color_spinor_field_order.h>
+#include <shared_memory_cache_helper.cuh>
+#include <madwf_transfer.h>
+#include <madwf_ml.h>
+
+/**
+  @file This file contains the argument and kernel that applies fifth dimension transfer to a vector:
+      out_s = T_{st} * in_t,
+  where s and t are fifth dimension indices. T_{st} is a spin matrix so T has shape Ls*4-by-Ls*4.
+*/
+
+namespace quda
+{
+
+  namespace madwf_ml
+  {
+
+    template <class real, int nSpin, int nColor, transfer_5D_t transfer_t> struct transfer_5D_mapper {
+    };
+
+    template <class real, int nSpin, int nColor> struct transfer_5D_mapper<real, nSpin, nColor, transfer_5D_t::Spin> {
+      using type = Matrix<complex<real>, nSpin>;
+    };
+
+    template <bool dagger, class real, int nSpin, int nColor>
+    __device__ __host__ inline void
+    matrix_vector_multiply(ColorSpinor<real, nColor, nSpin> &out, const Matrix<complex<real>, nSpin> &m, const ColorSpinor<real, nColor, nSpin> &v)
+    {
+#pragma unroll
+      for (int color = 0; color < nColor; color++) {
+#pragma unroll
+        for (int column = 0; column < nSpin; column++) {
+          auto v_col = v(column, color);
+#pragma unroll
+          for (int row = 0; row < nSpin; row++) {
+            if (dagger) {
+              out(row, color) = cmac(conj(m(column, row)), v_col, out(row, color));
+            } else {
+              out(row, color) = cmac(m(row, column), v_col, out(row, color));
+            }
+          }
+        }
+      }
+    }
+
+    template <class storage_t, int nSpin_, int nColor_, bool dagger_> struct Transfer5DArg : kernel_param<> {
+
+      static constexpr int nSpin = nSpin_;
+      static constexpr int nColor = nColor_;
+
+      using F = typename colorspinor_mapper<storage_t, 4, 3>::type;
```

Comment:
`4, 3` -> `nSpin, nColor`, or should we have an assert that `nSpin == 4`, `nColor == 3`?

Reply:
Good catch. The assert has already been performed in the corresponding `.cu` files.

Reply:
https://github.com/lattice/quda/pull/1061/commits/5cce447990c8ea085125244fea58154e4271dc8a